redash/query_runner/elasticsearch.py

```diff
@@ -139,27 +179,49 @@ def add_column_if_needed(mappings, column_name, friendly_name, result_columns, r
             for r in result_fields:
                 result_fields_index[r] = None
 
-        for h in raw_result["hits"]["hits"]:
-            row = {}
+        if 'error' in raw_result:
 
-            for field, column in ELASTICSEARCH_BUILTIN_FIELDS_MAPPING.iteritems():
-                if field in h:
-                    add_column_if_needed(mappings, field, column, result_columns, result_columns_index)
-                    row[column] = h[field]
+            error = raw_result['error']
+            if len(error) > 10240:
+                error = error[:10240] + '... continues'
 
-            column_name = "_source" if "_source" in h else "fields"
-            for column in h[column_name]:
-                if result_fields and column not in result_fields_index:
-                    continue
+            raise Exception(error)
 
-                add_column_if_needed(mappings, column, column, result_columns, result_columns_index)
+        elif 'aggregations' in raw_result:
 
-                value = h[column_name][column]
-                row[column] = value[0] if isinstance(value, list) and len(value) == 1 else value
+            if result_fields:
+                for field in result_fields:
+                    add_column_if_needed(mappings, field, field, result_columns, result_columns_index)
 
+            for key, data in raw_result["aggregations"].iteritems():
+                collect_aggregations(mappings, result_rows, key, data, None, result_columns, result_columns_index)
+
+            logger.debug("result_rows", str(result_rows))
+            logger.debug("result_columns", str(result_columns))
+
+        elif 'hits' in raw_result and 'hits' in raw_result['hits']:
+
+            if result_fields:
+                for field in result_fields:
+                    add_column_if_needed(mappings, field, field, result_columns, result_columns_index)
+
+            for h in raw_result["hits"]["hits"]:
+                row = {}
+
+                column_name = "_source" if "_source" in h else "fields"
+                for column in h[column_name]:
+                    if result_fields and column not in result_fields_index:
+                        continue
+
+                    add_column_if_needed(mappings, column, column, result_columns, result_columns_index)
+
+                    value = h[column_name][column]
+                    row[column] = value[0] if isinstance(value, list) and len(value) == 1 else value
 
-            if row and len(row) > 0:
                 result_rows.append(row)
+        else:
+
+            raise Exception('Seems you\'ve hit an unsupported feature')
```

Comment:
You can use "" here instead to avoid escaping (`raise Exception("Seems you've hit...")`).

Also considering this message will be relayed back to the user, let's rephrase it to something like:
"Redash failed to parse the results it got from ElasticSearch.", to make sure it's clear where the error is coming from.